### ISSUE TITLE
`trait_duplication_in_bounds` lint doesn't consider generics

### ISSUE THREAD

**Issue by aldhsu** [rust-lang/rust-clippy#8757]
### Summary

@giraffate caught this issue in https://github.com/rust-lang/rust-clippy/pull/8703 PR which is based on the TRAIT_DUPLICATION_IN_BOUNDS lint.

### Lint Name

TRAIT_DUPLICATION_IN_BOUNDS

### Reproducer

I tried this code:

```rust
trait BoundWithGeneric<T> {}

fn good_bounds_with_generic<T: BoundWithGeneric<u32>>(arg0: T, arg1: T)
where
    T: BoundWithGeneric<u64>,
{
    unimplemented!();
}

```

I saw this happen:

```
 --> $DIR/trait_duplication_in_bounds.rs:16:32
+   |
+LL | fn good_bounds_with_generic<T: BoundWithGeneric<u32>>(arg0: T, arg1: T)
+   |                                ^^^^^^^^^^^^^^^^^^^^^
+   |
+   = help: consider removing this trait bound

```

I expected to see this happen:

No linting error as while the same trait is used, the concrete type argument it is applied to is different.

### Version

```text
rustc 1.62.0-nightly (8f36334ca 2022-04-06)
binary: rustc
commit-hash: 8f36334ca939a67cce3f37f24953ff6f2d3f3d33
commit-date: 2022-04-06
host: x86_64-apple-darwin
release: 1.62.0-nightly
LLVM version: 14.0.0
```

**Comment by digama0** [comment]
Here are some more test cases for this issue:
```rust
#![warn(clippy::trait_duplication_in_bounds)]

trait Trait<A> {}
struct Foo;

// to avoid trivial bound lint
impl Trait<u8> for Foo {}
impl Trait<i8> for Foo {}
impl Trait<u8> for () {}
impl Trait<i8> for () {}

fn ok1() where Foo: Trait<u8> + Trait<i8> {}
fn bad2<A>() where Foo: Trait<u8> + Trait<i8> {}
fn ok3<A>() where (): Trait<u8> + Trait<i8> {}
fn bad3<A>() where Foo: Trait<A> + Trait<i8> {}
fn bad4<A, B>() where Foo: Trait<A> + Trait<B> {}
impl<A> Trait<(A,)> for Foo where Foo: Trait<A> + Trait<u8> {} // bad
impl<A, B> Trait<(A, B)> for Foo where Foo: Trait<A> + Trait<B> {} // bad

fn main() {}
```

**Comment by aldhsu** [comment]
I've been poking at this problem since it is related to my PR.

## Reproducer
I tried this code:
```rust
fn duplicate_trait<T: Clone + Clone + Default, Z: Copy>(arg0: T, arg1: Z) {}
```

I saw this happen:
```
error: this trait bound is already specified in the where clause
  --> $DIR/trait_duplication_in_bounds.rs:110:16
   |
LL | fn duplicate_trait<T: Clone + Clone + Default, Z: Copy>(arg0: T, arg1: Z) {}
   |                       ^^^^^
   |
   = help: consider removing this trait bound
```
The error message always says `this is trait bound is already specified in the where clause`. 

Two interpretations. 
1. The message is correct and the behaviour of the lint should only compare trait bounds to where clauses for the same generic, not within trait bounds or where clauses
2. The message is incorrect and behaviour is expected.
